buildtest/buildsystem/base.py

```diff
@@ -78,14 +82,6 @@ def load(self, config_file):
         """
         self.config_file = os.path.abspath(config_file)
 
-        if not os.path.exists(self.config_file):
```

Comment:
Is this check still done in self._validate_global()?

Reply:
this is not needed in ``self._validate_global()`` because it is done earlier in the ``discover_configs`` stage in particular at https://github.com/HPC-buildtest/buildtest-framework/blob/devel/buildtest/menu/build.py#L45 you will notice that we call discover_configs before calling BuildConfig so no point in doing these checks again

https://github.com/HPC-buildtest/buildtest-framework/blob/abcfad94b69598f87f15c58e3c0f0762d9c9a99f/buildtest/menu/build.py#L87

Reply:
Okay great! Just wanted to double check.

Reply:
Okay so there is an issue with Travis see https://travis-ci.com/github/HPC-buildtest/buildtest-framework/jobs/316447547 and it is related to how regtest was written since ``discover_config`` is not used in test before invoking BuildConfig

https://github.com/shahzebsiddiqui/buildtest-framework/blob/change_output_format/tests/buildsystem/test_base.py#L22

I guess either we refactor regtest or i undo my change. What your thoughts?

Reply:
The discover_config, in that it finds and validates a config file, must of course happen before you try to load it.

Reply:
Okay so i found that culprit. If i remove the lines https://github.com/shahzebsiddiqui/buildtest-framework/blob/change_output_format/tests/buildsystem/test_base.py#L21-22 then test pass. 

I see this test would be there when no config is passed and I think what you had in mind was if user runs ``buildtest build`` it should try to search from ``~/.buildtest/site`` i.e TESTROOT, but thing is we have a check at end if no config files found that it will exit which is here https://github.com/HPC-buildtest/buildtest-framework/blob/devel/buildtest/menu/build.py#L60

So i think it's not possible to have ``BuildConfig("")`` as a call since discover_config will catch all errors. Hope that makes sense. I ran test locally when commenting out line 21,22 and test was working. I was able to reproduce same error in Travis when the lines were uncommented.

So i'd recommend we should remove these lines.  I think there should be test for method ``discover_config`` which should catch all issues related to configuration, that way checks are done in advance before task is delegated to other classes. 
What do you think?

I think

Reply:
Also looking at codecov https://codecov.io/gh/HPC-buildtest/buildtest-framework/src/devel/buildtest/menu/build.py we dont have any test for ``discover_config`` and i think some regtest should be written to catch all types of input for configuration (file, directory, or path relative to ``.site``). That way the test are modular per method as pose to one or two monolothic test such as one in ``buildsystem/test_base.py``

Reply:
Okay travis is now  working with latest fix i proposed. I think this is the last remaining issue. I updated scope of PR as this will address three different issues. Let me know if this fix is acceptable and we can come to resolution.

Reply:
> Okay so i found that culprit. If i remove the lines https://github.com/shahzebsiddiqui/buildtest-framework/blob/change_output_format/tests/buildsystem/test_base.py#L21-22 then test pass.

This isn't a culprit - if the user doesn't provide a valid path the test should fail with systemExit, which was working before and is exactly what that test is doing. What seems to be the issue is that discover_config is not SystemExit-ing as it should. I think the lines should be kept, and the discover_config should system exit as it was originally done.

Reply:
Okay so would you want me to undo the change on line 81  and revert my regtest so that ``BuildTestConfig("")`` can be invoked which should raise SystemExit. 

```
if not os.path.exists(self.config_file):
```

Reply:
yes, remove the check that the config file exists isn't logical. The lines that I added for both a file, and then a directory, should not be removed.

Reply:
@vsoch i undid my changes and added test to validate ``discover_config`` See latest commit. This increases coverage of ``buildtest/menu/build.py`` which was previously at 20% now at 49%. See https://codecov.io/gh/HPC-buildtest/buildtest-framework/tree/devel/buildtest/menu

Reply:
I think you can see the coverage in coveralls here https://coveralls.io/builds/29955036